src/backend/cdb/cdbgroupingpaths.c

```diff
@@ -1551,6 +1554,66 @@ add_single_dqa_hash_agg_path(PlannerInfo *root,
 										ctx->agg_final_costs,
 										dNumGroups);
 		add_path(output_rel, path, root);
+
+		/*
+		 *  Finalize Aggregate
+		 *     -> Gather Motion
+		 *          -> Partial Aggregate
+		 *              -> HashAggregate, to remove duplicates
+		 *                          -> input
+		 */
+		path = (Path *) create_agg_path(root,
+										output_rel,
+										orig_path,
+										input_target,
+										AGG_HASHED,
+										AGGSPLIT_SIMPLE,
+										false, /* streaming */
+										dqa_group_clause,
+										NIL,
+										ctx->agg_partial_costs, /* FIXME */
+										clamp_row_est(dNumDistinctGroups / CdbPathLocus_NumSegments(distinct_locus)));
+
+		path = (Path *) create_agg_path(root,
+										output_rel,
+										path,
+										strip_aggdistinct(ctx->partial_grouping_target),
+										ctx->groupClause ? AGG_HASHED : AGG_PLAIN,
+										AGGSPLIT_INITIAL_SERIAL | AGGSPLITOP_DEDUPLICATED,
+										false, /* streaming */
+										ctx->groupClause,
+										NIL,
+										ctx->agg_partial_costs,
+										estimate_num_groups_on_segment(ctx->dNumGroupsTotal, input_rows, path->locus));
+		if (group_need_redistribute)
+			path = cdbpath_create_motion_path(root, path, NIL, false,
+										  group_locus);
+
+		path = (Path *) create_agg_path(root,
+										output_rel,
+										path,
+										ctx->target,
+										ctx->groupClause ? AGG_HASHED : AGG_PLAIN,
+										AGGSPLIT_FINAL_DESERIAL | AGGSPLITOP_DEDUPLICATED,
+										false, /* streaming */
+										ctx->groupClause,
+										ctx->havingQual,
+										ctx->agg_final_costs,
+										dNumGroups);
+
+		/*
+		 * Try disable other paths if users are eager this one.
+		 */
+		if (gp_eager_distinct_dedup)
```

Comment:
The prerequisites for getting the cheapest path are:
1. Very accurate statistical information;
2. A good cost estimation model for the optimizer;
If this cannot be achieved, GUC is OK to used for control